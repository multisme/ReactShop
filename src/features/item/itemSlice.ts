import { createSlice } from "@reduxjs/toolkit";
import { AppThunk } from "../../app/store";

export interface itemData {
  id: number | undefined;
  quantity: number;
  name: string;
  price?: number;
  details?: string;
  url?: string;
}

export const initialState = {
  id: undefined,
  quantity: 0,
  name: "unknown",
} as itemData;

const ItemSlice = createSlice({
  name: "item",
  initialState,
  reducers: {},
});
