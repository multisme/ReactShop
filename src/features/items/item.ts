import { createSlice } from "@reduxjs/toolkit";
import { AppThunk } from "../../app/store";

export interface itemData {
  id: number | undefined;
  quantity: number;
  name: string;
}

export const initialState = {
  id: undefined,
  quantity: 0,
  name: "unknown",
} as itemData;

const ItemSlice = createSlice({
  name: "item",
  initialState,
  reducers: {},
});
