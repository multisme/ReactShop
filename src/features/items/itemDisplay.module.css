.productList{
        list-style: none;
        display: flex;
}

.product{
        width: 150px;
        height: 150px;
}

.centeredContent{
        display: flex;
        align-items: center;
        justify-content: center;
}
