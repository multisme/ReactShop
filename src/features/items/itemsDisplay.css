.itemsList{
        list-style: none;
        display: flex;
        flex-wrap: wrap;
        margin: 0.25rem;
        padding: 0px;
}
