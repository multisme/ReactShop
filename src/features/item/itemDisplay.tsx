import React from "react";
import { useDispatch } from "react-redux";

import { itemData } from "features/item/itemSlice";
import { updateSelectedItem } from "features/showcase/showcaseSlice";

import "features/item/itemDisplay.css";

export const ItemDisplay = (item: itemData) => {
  //change the event type here with the correct one

  const dispatch = useDispatch();

  const selectItem = (e: any) => {
    e.preventDefault();
    dispatch(updateSelectedItem(item));
  };

  return (
    <li
      onClick={selectItem}
      className={"item flex-centered green-border"}
      key={item.id}
    >
      <div className="name">{item.name}</div>
      <div className={"quantity"}>{item.quantity}</div>
    </li>
  );
};
