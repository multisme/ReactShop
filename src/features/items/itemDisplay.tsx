import React from "react";
import {useDispatch} from "react-redux"

import { itemData } from "features/items/item";
import { updateSelectedItem } from "features/showcase/showcaseSlice"

import styles from 'features/items/itemDisplay.module.css';


interface Props{
        items: itemData[]
}


export const ItemListDisplay = ({ items }: Props) => {
        const renderedlist = items.map((item: itemData) => <ItemDisplay {...item}/>);
        return <ul className={styles.productList} id="list">{renderedlist}</ul>;
};

export const ItemDisplay = (item: itemData) => {
//change the event type here with the correct one

const dispatch = useDispatch();

const selectItem = (e: any) => {
        e.preventDefault();
       dispatch( updateSelectedItem(item) );
}

  return (
    <li
        onClick={selectItem}
        className={styles.product} key={item.id}
    >
      <div className="name">{item.name}</div>
      <div className={"quantity"}>{item.quantity}</div>
    </li>
  );
};
