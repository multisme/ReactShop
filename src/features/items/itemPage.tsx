import React from "react";
import { useDispatch, useSelector } from "react-redux";
import {Link} from "react-router-dom";
import {
        useHistory,
        useParams
 } from "react-router-dom";

import { RootState } from "app/rootReducer";
import {itemData} from "features/item/itemSlice";
import {itemPageSelector} from "features/items/itemsSlice";
import {addToCart} from "features/cart/cartSlice";

import { useFormFields } from "utils/utils";

interface productPageParams{
        id: string
}

const ItemPage = ()=>{
        const {id}  = useParams<productPageParams>();
        const [fields, handleFieldChange] = useFormFields({
                quantity: 0
        });
        const dispatch = useDispatch();
        const [item] = useSelector((state) =>itemPageSelector(state, id));
        const history = useHistory();
        if (id == undefined || item == undefined){
                return (
                <h3>Error</h3>
                );
        }
       
        const handleSubmit = (e: any) => {
                e.preventDefault();
                if (fields.quantity.content > 0){
                        dispatch(addToCart({
                        id: parseInt(id),
                        quantity: fields.quantity.content,
                        price: item.price}));
                        history.push("/bill")
                } else {
                        alert("Please choose something before checkout");
                }
        }
        
        const range = [...Array(item.quantity + 1).keys()]
        const availableQuantities = range.map((i) => 
                <option value={i} key={i}>{i}</option>
        )
 
        return (
        <div className="itemPage flex-centered">
                <div className="panel left flex-centered">
                <div className="mainImage">
                <img src="https://teamtijger.nl/wp-content/uploads/2018/06/zwarte-trui-transparant-300x300.png)" />;
                </div>
                </div>
                <div className="panel right">
                        <div className="info">
                                <div className="name">
                                {item.name}
                                </div>
                                <div className="price">
                                {item.price}
                                </div>
                                <div className="details">
                                {item.details}
                                </div>
                        </div>
                <form className="itemForm" onSubmit={handleSubmit}>
                <fieldset className="quantity">
                <select name="quantity" id="quantity" onChange={handleFieldChange}>
                        {availableQuantities}
                </select>
                </fieldset>
                <fieldset className="button">
                <Link to="/">
                <button id="goBack" className="buy" >
                        GO BACK
                </button>
                </Link>
                <button type="submit" id="checkout" className="checkout">
                        CHECKOUT
                </button>
                </fieldset>
                </form>
                </div>
        </div>
        )
}
export default ItemPage;
