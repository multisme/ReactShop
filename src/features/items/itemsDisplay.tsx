import React from "react";

import {useDispatch} from "react-redux"
import {Link} from "react-router-dom"

import { itemData } from "features/item/itemSlice";
import {
        updateSelectedItem,
        removeSelectedItem
} from "features/showcase/showcaseSlice";

import "features/items/itemsDisplay.css";

interface itemListProps{
items: itemData[]
}

interface itemProps{
item: itemData
}

const iconPath = process.env.PUBLIC_URL + '/assets/images/';

console.log("cin", iconPath);
export const ItemListDisplay = ({ items }: itemListProps) => {
        const renderedlist = items.map((item: itemData) =>
                        <div className="itemCard"  key={item.id} style={{backgroundImage: "url("  +item.url +")"}}>
                        <Link key={item.id}to={`/products/${item.id}`}>
                        <ItemDisplay item={item}/>
                        </Link>
                        </div>
                        );
        return (
                        <ul className="itemsList" id="list">
                        {renderedlist}
                        </ul>
               );
};

export const ItemDisplay = ({item}: itemProps) => {
        //change the event type here with the correct one

        const dispatch = useDispatch();

        const selectItem = (e: any) => {
                console.log("mouse enter");
                e.preventDefault();
                dispatch( updateSelectedItem(item) );
        }

        const unselectItem = (e: any) => {
                dispatch( removeSelectedItem() );
        }

        return (
                        <li
                        onMouseEnter={selectItem}
                        onMouseLeave={unselectItem}
                        className={"flex-centered"}
                        >
                        <div className="picture">
                        </div>
                        <div className={"detailsSmall flex-centered"}>
                        <div className="name">{item.name}</div>
                        <div className={"price"}>{item.price}â‚¬</div>
                        </div>
                        </li>
               );
};
