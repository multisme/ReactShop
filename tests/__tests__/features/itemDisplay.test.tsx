import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';

import { ItemListDisplay } from "features/items/itemsDisplay";
import { itemData } from "features/item/itemSlice";

const mockStore = configureStore();

describe('itemDisplay', () => {
        let store: any;
 
        beforeEach(() => {
                store = mockStore({
                myState: 'sample text',
        });
        });

        it("gets an empty array", () => {
                const items: itemData[] = [];
                const component= renderer.create(
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                ).toJSON();
                expect(component).toMatchSnapshot();
}) 

describe('ItemListDisplay', () => {
        //Check correct type
        let store: any;
 
        beforeEach(() => {
                store = mockStore({
                myState: 'sample text',
        });
        });

        it("gets an empty array", () => {
                const items: itemData[] = [];
                const component= renderer.create(
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                ).toJSON();
                expect(component).toMatchSnapshot();
        });
        it("gets array of one", () => {
                const items: itemData[] = [
                        {"id": 3, "name": "testname", "quantity": 4},
                ];
                const component= renderer.create(
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                ).toJSON();
                expect(component).toMatchSnapshot();
        });
        it("gets array of some", () => {
                const items: itemData[] = [
                        {"id": 3, "name": "testname", "quantity": 4},
                        {"id": 4, "name": "postname", "quantity": 8}
                ];
                const component= renderer.create(
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                ).toJSON();
                expect(component).toMatchSnapshot();
        })
})
