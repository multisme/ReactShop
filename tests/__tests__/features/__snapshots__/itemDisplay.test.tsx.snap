// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemListDisplay gets an empty array 1`] = `
<ul
  className="itemsList"
  id="list"
/>
`;

exports[`ItemListDisplay gets array of one 1`] = `
<ul
  className="itemsList"
  id="list"
>
  <li
    className="item flex-centered green-border"
    onClick={[Function]}
  >
    <div
      className="name"
    >
      testname
    </div>
    <div
      className="quantity"
    >
      4
    </div>
  </li>
</ul>
`;

exports[`ItemListDisplay gets array of some 1`] = `
<ul
  className="itemsList"
  id="list"
>
  <li
    className="item flex-centered green-border"
    onClick={[Function]}
  >
    <div
      className="name"
    >
      testname
    </div>
    <div
      className="quantity"
    >
      4
    </div>
  </li>
  <li
    className="item flex-centered green-border"
    onClick={[Function]}
  >
    <div
      className="name"
    >
      postname
    </div>
    <div
      className="quantity"
    >
      8
    </div>
  </li>
</ul>
`;
